core/src/main/java/com/linecorp/armeria/server/RedirectService.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+
+/**
+ * A {@link Service} that implements an HTTP redirection.
+ *
+ * <p>This class will redirect requests from its bound {@link Service} path pattern
+ * to a new location path pattern. Currently only these path patterns are supported:
+ * <ul>
+ *   <li>{@code /new} (no path parameters)</li>
+ *   <li>{@code /new/{var1}} (curly-brace style)</li>
+ *   <li>{@code /new/:var1/new/:var2} (colon style)</li>
+ *   <li>{@code http://host/new} (full url no path parameters)</li>
+ *   <li>{@code http://host/new/{var1}} (full url and curly-brace style param)</li>
+ *   <li>{@code http://host/new/:var1/new/:var2} (full url and colon style param)</li>
+ * </ul>
+ * The {@link RedirectService} will return {@link HttpStatus#TEMPORARY_REDIRECT 307 Temporary Redirect}
+ * by default.
+ * <pre>{@code
+ * ServerBuilder sb = ...;
+ * sb.service("/old/", new RedirectService("/new"));
+ * sb.service("/old/{var}", new RedirectService("/new/{var}");
+ * sb.service("/old/{var}", new RedirectService(ctx -> "/new/" + ctx.pathParam("var"));
+ *
+ * sb.service("/old/{var}", new RedirectService(HttpStatus.MOVED_PERMANENTLY, "/new/{var}");
+ * sb.service("/old/{var}", new RedirectService("http://user:name@localhost:8080/search/{var}");
+ * }</pre>
+ */
+public class RedirectService extends AbstractHttpService {
+
+    private static final Pattern VALID_DEFAULT_URI_PATTERN = Pattern
+            .compile("(((http|https)://).*(:[0-9]+)?)?(?:/[^/{}:]+|/:[^/{}]+|/\\{[^/{}]+})+/?");
```

Comment:
Think we can optimize this regex but guess it would be ok to keep it as is since the code will not be run on critical pass, right?
e.g. 1) you can avoid captures, 2) at `[0-9]+`, you can specify [quantifiers](https://docs.oracle.com/javase/tutorial/essential/regex/quant.html)

Reply:
added range quantifiers. however still will allow invalid port values though
`http://host:99999/`